### PR TITLE
Cleanup dumping/to_dict

### PR DESCRIPTION
**Pull request by alejoe91**
This PR simplifies a bit the dumping and cleans up some to_dict mechanisms.

In brief:

 - the `dump_to_json` is always `recursive` because that allows to modify the paths in the dict. The `SIJsonEncoder` basically does the same but at dump time
 - the `dump_to_pickle` is always not recursive and it doesn't allow `relative_to` to be set (because it's incompatible with the pickle loading and it would require in-place object modifications)

### CODE REVIEW

File: src/spikeinterface/core/base.py [hunk 1]
```diff
@@ -338,6 +338,9 @@ def to_dict(
 
         kwargs = self._kwargs
 
+        if relative_to and not recursive:
+            raise Exception("`relative_to` is only posible when `recursive=True`")
```
Comment (h-mayorquin) [inline src/spikeinterface/core/base.py:342]: Yeah, I agree with @DradeAW that it should be a ValueError.

File: src/spikeinterface/core/base.py [hunk 2]
```diff
@@ -561,23 +564,28 @@ def dump(self, file_path: Union[str, Path], relative_to=None, folder_metadata=No
...
-            This means that file and folder paths in extractor objects kwargs are changed to be relative rather than absolute.
+        relative_to: str, Path, or None
+            If not None, file_paths are serialized relative to this path
+        folder_metadata: str, Path, or None
+            Folder with files containing additional information (e.g. probe in BaseRecording) and properties.
         """
-        assert self.check_if_dumpable()
+        assert self.check_if_json_serializable(), "The extractor is not json serializable"
 
         if relative_to is True:
```
Comment (h-mayorquin) [inline src/spikeinterface/core/base.py:580]:
```suggestion
        # Writing paths as relative_to requires recursively expanding the dict
        if relative_to:
           recursive = True
           # We use relative_to == True to encode using the parent_folder
           relative_to = Path(file_path).parent if relative_to is True else relative_to
```

What do you think?

File: src/spikeinterface/core/base.py [hunk 3]
```diff
@@ -561,23 +564,28 @@ def dump(self, file_path: Union[str, Path], relative_to=None, folder_metadata=No
...
+        The extractor can be re-loaded with load_extractor(json_file)
 
         Parameters
         ----------
         file_path: str
             Path of the json file
-        relative_to: str, Path, True or None
-            If not None, files and folders are serialized relative to this path. If True, the relative folder is the parent folder.
-            This means that file and folder paths in extractor objects kwargs are changed to be relative rather than absolute.
+        relative_to: str, Path, or None
```
Comment (h-mayorquin) [inline src/spikeinterface/core/base.py:573]: Why did you modify the docstring to the old typing?

**Review by h-mayorquin — APPROVED** [review]
This looks good to me. Minor questions about documentation.

File: src/spikeinterface/core/base.py (hunk 3 above)
Comment (alejoe91) [reply src/spikeinterface/core/base.py:573]: mistake in merging, fixing back

**Review by samuelgarcia — APPROVED** [review]

**Pull request merged**